lib/mapping-transform.js

```diff
@@ -11,28 +11,22 @@ module.exports = class MappingTransform extends Transform {
     }
 
     _transform(chunk, _, callback) {
+        let value;
```

Comment:
Life is easier if you pick with a sync or async style and stick with it. Mixing the two is a common source of bugs. 

For example, if `this._function` throws you're emitting an error on the current stream. However, if `this._function` returns a rejected promise, the callback is called with an error. This is wrong. Throwing and returning a rejected promise are semantically identical. The only difference is that one of them happens synchronously and the other async. If something about this distinction as implemented is important, we have bigger latent bugs here.

Considering `this._function` is possibly async and `_transform` cannot be async, sticking with a promise chain is your best bet.

```suggestion
    _transform(chunk, _, callback) {
        Promise.resolve(chunk)
            .then(req => this._function(req))
            .then(res => callback(null, res))
            .catch(err => callback(new RiffError(errorTypes.REQUEST_REPLY_FUNCTION_RUNTIME, err));
    }
```

Reply:
It does not seem to work.
Using the adapter and the function described [here](https://github.com/projectriff/node-function-invoker/issues/133#issuecomment-584075817), the request hangs.

Reply:
It seems that the request does _not_ hang when using `this.emit('error', error)` in the catch clause of the try-catch block surrounding `this._function`.
I agree this does seem right. 

Not sure why `callback(error)` does not have the same effect, nor `this.emit('error', ...)` in a promise chain.

Reply:
I think we need to stick to a synchronous style, else `_transform` will return before the promise chain completes.

Reply:
Note that the current PR contents fix the problem with synchronous functions but the same problem still occurs with async functions.

I am not sure we can really reconcile the synchronous nature of streams and the possibly async nature of functions here.

Reply:
Thanks to https://github.com/projectriff/node-function-invoker/pull/147, I'm now much more familar with promises than before.
@scothis your initial suggestion works well, the only issue with it is that `callback(someError)` throws in the promise `catch`. Because of that, the promise ends up being rejected and given there is no subsequent `catch`, then the warning appears.

Reply:
FYI: I just asked [nodejs/help](https://github.com/nodejs/help/issues/2477) to see what's the best course of action.

Reply:
Seems `process.nextTick` is the way to go.